lib/puppet/parser/scope.rb

```diff
@@ -174,6 +173,84 @@ def add_entries_to(target = {})
 
   end
 
+  class ParameterScope < Ephemeral
+    class Access
+      attr_accessor :value
+
+      def assigned?
+        instance_variable_defined?(:@value)
+      end
+    end
+
+    # A parameter default must be evaluated using a special scope. The scope that is given to this method must
+    # have a `ParameterScope` as its last ephemeral scope. This method will then push a `MatchScope` while the
+    # given `expression` is evaluated. The method will catch any throw of `:unevaluated_parameter` and produce
+    # an error saying that the evaluated parameter X tries to access the unevaluated parameter Y.
+    #
+    # @param name [String] the name of the currently evaluated parameter
+    # @param expression [Puppet::Parser::AST] the expression to evaluate
+    # @param scope [Puppet::Parser::Scope] a scope where a `ParameterScope` has been pushed
+    # @return [Object] the result of the evaluation
+    #
+    # @api private
+    def self.evaluate3x(name, expression, scope)
+      elevel = scope.ephemeral_level
+      bad = catch(:unevaluated_parameter) do
+        scope.new_match_scope(nil)
+        return expression.safeevaluate(scope)
+      end
+      raise Puppet::Error, "$#{name} tries to access unevaluated $#{bad}"
+    ensure
+      scope.unset_ephemeral_var(elevel)
+    end
+
+    def self.evaluate(name, expression, scope, evaluator)
+      elevel = scope.ephemeral_level
+      bad = catch(:unevaluated_parameter) do
+        scope.new_match_scope(nil)
+        return evaluator.evaluate(expression, scope)
+      end
+      raise Puppet::Error, "$#{name} tries to access unevaluated $#{bad}"
+    ensure
+      scope.unset_ephemeral_var(elevel)
+    end
+
+    def initialize(parent, param_names)
+      super(parent)
+      @params = {}
+      param_names.each { |name| @params[name] = Access.new }
+    end
+
+    def [](name)
+      access = @params[name]
+      return super if access.nil?
+      throw(:unevaluated_parameter, name) unless access.assigned?
+      access.value
+    end
+
+    def []=(name, value)
+      @params[name] ||= Access.new
```

Comment:
what if user writes something confused like some variation of this:
```
function foo($x = ($y = 2), $y = 5) { }
```
It does not introduce new variables, but assigns to parameters. I am asking because if this method is called and it is not a parameter I think it should error.

Remember that if default expression calls functions, the scope will potentially be accessed by 3d party ruby code (doing odd things they are not allowed to).